gymnasium/envs/mujoco/mujoco_env_base.py

```diff
@@ -0,0 +1,420 @@
+        raise NotImplementedError
+
+    def reset_model(self) -> NDArray[np.float64]:
+        """
+        Reset the robot degrees of freedom (qpos and qvel).
+        Implement this in each subclass.
+        """
+        raise NotImplementedError
+
+    def _initialize_simulation(self) -> Tuple[Any, Any]:
+        """
+        Initialize MuJoCo simulation data structures mjModel and mjData.
+        """
+        raise NotImplementedError
+
+    def _reset_simulation(self) -> None:
+        """
+        Reset MuJoCo simulation data structures, mjModel and mjData.
+        """
+        raise NotImplementedError
+
+    def _step_mujoco_simulation(self, ctrl, n_frames) -> None:
+        """
+        Step over the MuJoCo simulation.
+        """
+        raise NotImplementedError
+
+    def render(self) -> Union[NDArray[np.float64], None]:
+        """
+        Render a frame from the MuJoCo simulation as specified by the render_mode.
+        """
+        raise NotImplementedError
+
+    # -----------------------------
+    def _get_reset_info(self) -> Dict[str, float]:
+        """Function that generates the `info` that is returned during a `reset()`."""
+        return {}
+
+    def reset(
+        self,
+        *,
+        seed: Optional[int] = None,
+        options: Optional[dict] = None,
+    ):
+        super().reset(seed=seed)
+
+        self._reset_simulation()
+
+        ob = self.reset_model()
+        info = self._get_reset_info()
+
+        if self.render_mode == "human":
+            self.render()
+        return ob, info
+
+    def set_state(self, qpos, qvel) -> None:
+        """
+        Set the joints position qpos and velocity qvel of the model. Override this method depending on the MuJoCo bindings used.
+        """
+        assert qpos.shape == (self.model.nq,) and qvel.shape == (self.model.nv,)
+
+    @property
+    def dt(self) -> float:
+        return self.model.opt.timestep * self.frame_skip
+
+    def do_simulation(self, ctrl, n_frames) -> None:
+        """
+        Step the simulation n number of frames and applying a control action.
+        """
+        # Check control input is contained in the action space
+        if np.array(ctrl).shape != (self.model.nu,):
+            raise ValueError(
+                f"Action dimension mismatch. Expected {(self.model.nu,)}, found {np.array(ctrl).shape}"
+            )
+        self._step_mujoco_simulation(ctrl, n_frames)
+
+    def close(self):
+        """Close all processes like rendering contexts"""
+        raise NotImplementedError
+
+    def get_body_com(self, body_name) -> NDArray[np.float64]:
+        """Return the cartesian position of a body frame"""
+        raise NotImplementedError
+
+    def state_vector(self) -> NDArray[np.float64]:
+        """Return the position and velocity joint states of the model"""
+        return np.concatenate([self.data.qpos.flat, self.data.qvel.flat])
+
+
+class MuJocoPyEnv(BaseMujocoEnv):
```

Comment:
Why is this here? (it is also in `mujoco_py_env.py`)

Reply:
Was an accident, deleted it